hoodie-common/src/main/java/com/uber/hoodie/common/util/collection/DiskBasedMap.java

```diff
@@ -106,7 +106,16 @@ private void initFile(File writeOnlyFileHandle) throws IOException {
     if (!writeOnlyFileHandle.getParentFile().exists()) {
       writeOnlyFileHandle.getParentFile().mkdir();
     }
-    writeOnlyFileHandle.createNewFile();
+    // create the file
+    try {
+      writeOnlyFileHandle.createNewFile();
```

Comment:
This does not make sense to me. If the create failed, why would we catch and then attempt to delete it ? 
You can fix this in the following way : 

There is already a shutdownHook called -> addShutDownHook(..) in this class. Just add the following line there : 

writeOnlyFileHandle.delete().

Reply:
I think the issue I had was caused by this line. The job failed to create the tmp file and died in the halfway, so the incomplete tmp file was left behind. It didn't have the chance to reach `writeOnlyFileHandle.delete()` and `addShutDownHook()` below. The failure point is https://github.com/apache/incubator-hudi/pull/795/files#diff-673dd1309d6ea3874ce463a67eeae36cR228. You could also check with the stack trace I attached on the JIRA ticket.

Reply:
The shutdownHook() is called just before the JVM exists and is controlled differently than what you see in the code. So if you add it there, the JVM exit should clear/delete the file.

Reply:
gotcha, will update the code later today

Reply:
Not sure even the shutdown hook is fool proof here.. it wont be called if the jvm is kill -9ed by YARN for e.g   can we look into registering each file we create for deleteOnExit()?

Reply:
@vinothchandar sorry I don't really understand what do you mean by registering, would you explain a little bit more?

Reply:
Something like this https://www.mkyong.com/java/how-to-delete-temporary-file-in-java/

Reply:
@vinothchandar The delteOnExit() code is already present, that's how we clean up all files here. In this case, what @garyli1019 is seeing is that the job fails right after creation before the deleteOnExit(). Either the deleteOnExit is moved to finally or we register shutdown with the exception when sigkill -9 is invoked there may be files lying around.

Reply:
I have 0 java experience but based on the code I read, the tmp file was created first then the job use java outputSteam to fill the file, https://github.com/apache/incubator-hudi/blob/master/hoodie-common/src/main/java/com/uber/hoodie/common/util/SpillableMapUtils.java#L86. So trying to catch an exception in the file creation in this PR won't work since the file was successfully created and already registered to delete. I probably need to catch an exception somewhere else.

Reply:
@garyli1019 the best approach is to leave the deleteOnExit() as is and add a delete in the shutdown hook as discussed (that's best effort).

Reply:
@n3nash The `writeOnlyFileHandle` in `addShutDownHook` is `sizeAwareDataOutputStream` type instead of `File` so it doesn't have a `delete()` method. In `initFile()`, `writeOnlyFileHandle` is a `File` type. Should we add a variable to this class in order to call it in `addShutDownHook`?

Reply:
sure, see what makes sense, you could remove the filePath and just keep a handle to the file in the class.